beacon_node/lighthouse_network/src/discovery/mod.rs

```diff
@@ -1036,12 +1038,107 @@ impl<TSpec: EthSpec> NetworkBehaviour for Discovery<TSpec> {
             FromSwarm::DialFailure(DialFailure { peer_id, error, .. }) => {
                 self.on_dial_failure(peer_id, error)
             }
+            FromSwarm::NewListenAddr(ev) => {
+                /* TODO(jmcph4): plumb user override status into config somehow */
+                let addr: &Multiaddr = ev.addr;
+                let listener_id: ListenerId = ev.listener_id;
+
+                trace!(self.log, "Received NewListenAddr event from swarm"; "listener_id" => ?listener_id, "addr" => ?addr);
+
+                let ip: IpAddr = match addr.iter().next() {
+                    Some(Protocol::Ip4(ip)) => IpAddr::V4(ip),
+                    Some(Protocol::Ip6(ip)) => IpAddr::V6(ip),
+                    _ => {
+                        debug!(self.log, "Encountered unacceptable multiaddr for listening (no IP)"; "addr" => ?addr);
+                        return;
+                    }
+                };
+
+                let update_enr = |ip: IpAddr, transport: Protocol| {
+                    /* this closure shouldn't exist but `Ipv4Addr::is_global` is Nightly at the moment */
+                    let is_global = |ip: Ipv4Addr| {
+                        !(ip.is_broadcast()
+                            || ip.is_documentation()
+                            || ip.is_link_local()
+                            || ip.is_loopback()
+                            || ip.is_multicast()
+                            || ip.is_private()
+                            || ip.is_unspecified())
+                    };
+
+                    let insert_into_enr = |key: &str, port: &u16| {
+                        if let Err(e) = self.discv5.enr_insert(key, port) {
+                            warn!(self.log, "Failed to write to ENR"; "error" => ?e);
+                        }
+                    };
+
+                    let mut should_update_ip: bool = true;
+
+                    /* we cannot violate the ENR invariant */
+                    if let IpAddr::V4(ip4) = ip {
+                        if !is_global(ip4) {
+                            debug!(self.log, "Refusing to update ENR IP (unreachable)"; "ip" => ?ip);
+                            should_update_ip = false;
+                        }
+                    }
+
+                    let port: u16 = match transport {
+                        Protocol::Tcp(port) => port,
+                        Protocol::Udp(port) => port,
+                        _ => {
+                            crit!(self.log, "Attempt to update discovery with unsupported transport (this should never occur)"; "transport" => ?transport);
```

Comment:
I think on every start we now also listen on the QUICv1 protocol. 

I would imagine that means running this code would immediately emit this crit. 

I usually run an instance of lighthouse locally (with backfill disabled and weak subjectivity sync) to see if it starts up and connects and doesn't emit crits. 

It's usually a good self-test to catch bugs like this.

Reply:
> I think on every start we now also listen on the QUICv1 protocol.

As of #4402 this is correct.

> I would imagine that means running this code would immediately emit this crit.

When we listen on QUICv1, the multiaddress looks like this: `/ip4/127.0.0.1/udp/9001/quic-v1`/. In this handler (for `FromSwarm::NewListenAddr`), we parse the multiaddress from left-to-right and only ever two `Protocol`s deep -- so we'll never encounter this `crit!` log emission during normal operation of the BN. This is still a bug obviously as only ever parsing two `Protocol`s deep will completely miss QUICv1 altogether.

Reply:
Oh right. I saw this line: 
https://github.com/jmcph4/lighthouse/blob/10263d3e003218c529f420d7e34ad658ff2f1ca1/beacon_node/lighthouse_network/src/discovery/mod.rs#L1131
```rust
                match addr.iter().nth(1) {
                    Some(Protocol::Tcp(port)) => update_enr(ip, Protocol::Tcp(port)),
                    Some(Protocol::Udp(port)) => update_enr(ip, Protocol::Udp(port)),
                    Some(Protocol::Quic) => update_enr(ip, Protocol::Quic),
                    _ => {
                        debug!(self.log, "Encountered unacceptable multiaddr for listening (invalid transport)"; "addr" => ?addr);
                    }
                };
```
which explicitly passes the Quic protocol to that function. I guess this case just never occurs because you only look at the nth(1).

Reply:
Yeah, it's somewhat regrettable that the current fix (i.e., 599c706e673b4ba90cf9c600227049c11930e146) duplicates the port number itself (as the `Protocol::QuicV1` variant doesn't store the port unlike its TCP and UDP counterparts) but is much more explicit about the mapping from multiaddress to ENR update information.